### PR TITLE
feat(hooks): Create useRepositories hook

### PR DESCRIPTION
**Pull request by NisanthanNanthakumar**
## Objective:
Creates the useRepositories hook. This should work with our existing RepositoryStore in Reflux. I refactored the withRepositories HOC to use the useRepositories hook internally to minimize duplicate logic. The hook handles fetching, pagination, error handling, and other niceties.

### CODE REVIEW

File: static/app/utils/useRequest.tsx [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    query?: string;
+    cursor?: typeof cursor;
+    per_page?: number;
+  } = {};
+
+  if (search) {
+    query.query = `${query.query ?? ''} ${search}`.trim();
+  }
+
+  const isSameSearch = lastSearch === search || (!lastSearch && !search);
```
Comment (leeandher) [inline static/app/utils/useRequest.tsx:82]: `isSameSearch` will be `true` if the it compares `''` and `null`. Not sure if this intended just FYI.

File: static/app/utils/useRequest.tsx [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+  nextPage: () => Promise<void>;
+  get: () => Promise<void>;
+  post: (_) => Promise<void>;
+  update: (_) => Promise<void>;
+  del: () => Promise<void>;
+} & State;
+
+type Options = {
+  /**
+   * Number of repositories to return
```
Comment (leeandher) [inline static/app/utils/useRequest.tsx:52]: I believe this comment should be generic, unless `useRequest` will only be used for `useRepositories`

File: static/app/utils/useRequest.tsx [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+type FetchRequestOptions = {
+  limit?: Options['limit'];
+  cursor?: State['nextCursor'];
+  search?: State['lastSearch'];
+  lastSearch?: State['lastSearch'];
+};
+
+/**
+ * Helper function to actually load repositories
```
Comment (leeandher) [inline static/app/utils/useRequest.tsx:65]: Same as above, should be generic

File: static/app/utils/useRepositories.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  if (store.orgSlug && store.orgSlug !== orgId) {
+    RepositoryActions.resetRepositories();
+  }
+
+  const {get, ...rest} = useRequest(
+    `/organizations/${orgId}/repos/`,
+    {
+      limit,
+    },
```
Comment (leeandher) [inline static/app/utils/useRepositories.tsx:41]: super-nit: Remove the comma to have it flow to one line only

File: static/app/utils/useRequest.tsx [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  }
+
+  const isSameSearch = lastSearch === search || (!lastSearch && !search);
+
+  if (isSameSearch && cursor) {
+    query.cursor = cursor;
+  }
+
+  if (limit !== undefined) {
+    query.per_page = limit;
```
Comment (leeandher) [inline static/app/utils/useRequest.tsx:89]: Is this accurate? How would I specify a maximum of 20 items, but 5 per page?

File: static/app/utils/useRequest.tsx [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+    if (search === '') {
+      return;
+    }
+
+    if (orgId === undefined) {
+      // eslint-disable-next-line no-console
+      console.error(
+        'Cannot use useRepositories.onSearch without an organization in context'
+      );
```
Comment (leeandher) [inline static/app/utils/useRequest.tsx:163]: Should this be here?  Can we capture the error in Sentry instead?

File: static/app/utils/useRequest.tsx [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+      onSuccessCb(results);
+      setState({
+        ...state,
+        data: results,
+        hasMore,
+        loading: false,
+        nextCursor,
+      });
+    } catch (err) {
+      console.error(err); // eslint-disable-line no-console
```
Comment (leeandher) [inline static/app/utils/useRequest.tsx:145]: See below, I think we should be capturing these errors in Sentry

File: static/app/stores/repositoryStore.tsx [hunk 8]
```diff
@@ -79,6 +76,10 @@ const storeConfig: Reflux.StoreDefinition & RepositoryStoreInterface = {
   get() {
     return {...this.state};
```
Comment (leeandher) [inline static/app/stores/repositoryStore.tsx:77]: nit: I know this isn't your change but this operation seems redundant

File: static/app/utils/useRepositories.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    data => RepositoryActions.loadRepositoriesSuccess(data)
+  );
+
+  useEffect(() => {
+    /** Multiple components may call useRepositories.
+     * We want to prevent making multiple calls for the same org. */
```
Comment (evanpurkhiser) [inline static/app/utils/useRepositories.tsx:47]: Just write this comment with `//`s lol

File: static/app/utils/useRequest.tsx [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  if (pageLinks) {
+    const paginationObject = parseLinkHeader(pageLinks);
+    hasMore = paginationObject?.next?.results || paginationObject?.previous?.results;
+    nextCursor = paginationObject?.next?.cursor;
+  }
+
+  return {results: data, hasMore, nextCursor};
+}
+
+function useRequest(
```
Comment (evanpurkhiser) [inline static/app/utils/useRequest.tsx:109]: Can you please open this as it's own PR so we can review it away from useRepositories?

File: static/app/stores/repositoryStore.tsx [hunk 11]
```diff
@@ -79,6 +76,10 @@ const storeConfig: Reflux.StoreDefinition & RepositoryStoreInterface = {
...
   },
+
+  getState() {
+    return this.get();
+  },
```
Comment (evanpurkhiser) [inline static/app/stores/repositoryStore.tsx:82]: Could you open this as a separate PR `ref(js): Make repositoryStore useLegacyStore compatible`?

**Review by evanpurkhiser — CHANGES_REQUESTED** [review]

File: static/app/utils/useRequest.tsx (hunk 5 above)
Comment (NisanthanNanthakumar) [reply static/app/utils/useRequest.tsx:89]: @leeandher hmm our endpoints don't currently support `maximum # of results returned`. We only support per page limits.

File: static/app/utils/useRequest.tsx (hunk 1 above)
Comment (NisanthanNanthakumar) [reply static/app/utils/useRequest.tsx:82]: @leeandher I think that's fine. We don't need to make new requests for empty search strings.

**Comment by mgaeta** [comment]
Can I close this?

**Pull request closed without merging**